### ISSUE TITLE
🦔 Investigate switching to urql Graphcache in order to use optimistic updates

### ISSUE THREAD

**Issue by tristan-orourke** [GCTC-NTGC/gc-digital-talent#8309]
## Background

When you save an inline form, it immediately goes back to the non-editable version, but showing the old data. Only when the request comes back successfully does the content update to what you just submitted. That delay can be a bit jarring, especially if it takes longer than a second. 

This ticket was originally looking for any solution to the UI delay. I've converted that into a new ticket, #9515, and made this one more specific to one solution we looked into.

## 🦋 Expected Behaviour

When you edit data in a form and submit, the app should immediately treat that data as true, unless/until the request fails.

This is exactly the promise of URQL's optimistic update feature! 
https://formidable.com/open-source/urql/docs/graphcache/cache-updates/#optimistic-updates

## 🕵️ Details

The main prerequisite to using the optimistic update feature is switching to urql graphecache -  we currently use the simpler and more naive document cache.

We don't fully understand yet the implications of this switch, but we assume they will be significant. The optimisic updates _could_ mean a significant overall improvement to UX, but we need to investigate and test the use of graphecache before we can confidently switch.

## Questions

- What are the risks of using graphecache instead of document cache?
- What we need to do to make graphecache work for our needs? How much new or changed code would it require?
- If we can switch to graphecache, how much new or changed code would Optimistic Updates require?
- Can it properly handle polymorphic relationships in graphql?  What are the limitations. 

Timebox: 2-3 days

**Comment by gobyrne** [comment]
Optimistic updating using the data that was collected could solve this. In the event of an error, the data reverts.

**Comment by esizer** [comment]
Looks like the optimistic updates should work. However, the documentation makes it seem like it requires using the graph cache, which requires urql@4.0.0, which we can't upgrade too until #7887 is done 😢 

Unless we consider this higher priority and want to install the previous version until we get a chance to circle back and bump auth, core and cache. Though, that does seem like some unnecessary effort even if [the changes are not significant](https://github.com/urql-graphql/urql/releases/tag/%40urql%2Fexchange-graphcache%406.0.0).

**Comment by vd1992** [comment]
I don't think this is high priority so waiting until the dependencies are sorted would be best. Especially since it could make things messier down the road like reverting the auth bump revert.

**Comment by esizer** [comment]
> I don't think this is high priority so waiting until the dependencies are sorted would be best. Especially since it could make things messier down the road like reverting the auth bump revert.

I attempted to do this on v5 instead of v6 and migrating to the graphcache seemed to be more difficult than their documentation made it seem 😬 

> This will automatically enable normalized caching, and you may find that in a lot of cases, Graphcache already does what you'd expect it to do without any additional configuration.

Not for us, apparently 😢 

REF: [Graphcache installation and setup](https://formidable.com/open-source/urql/docs/graphcache/#:~:text=This%20will%20automatically%20enable%20normalized%20caching%2C%20and%20you%20may%20find%20that%20in%20a%20lot%20of%20cases%2C%20Graphcache%20already%20does%20what%20you%27d%20expect%20it%20to%20do%20without%20any%20additional%20configuration.)

**Comment by brindasasi** [comment]
Team decided to go ahead with spinner solution and implement Graphcache implementation as a long term solution for later.

**Comment by esizer** [comment]
I tried again with a spinner but it doesn't work since we have no awareness of if the cache is updating. So as far as we know, the data has been updated as soon as we return to the display mode. Unfortunately, I think we will need to migrate to the graphcache and use optimistic updates.

So. I tried to graphcache but it is having trouble with how our experience types extend interfaces. We may need to wait until we can upgrade these libraries. Some issues we had:

- Introspection is weird because our version is behind
- Experience types need introspection because they extend interfaces
- It seems graphcache wants mutations to return the fields they modify so the cache can be updated

We may need to look into our queires/mutations to only return the data we need from them and nothing more/less to get the most out of the graphcache.

**Comment by vd1992** [comment]
To make a plan to go forward, what would you call the main blockers then for moving forward. As in, of the things you noted what should be spun off as issues marked as blocking this. If we can't progress on this issue itself we can at least plan the next steps

**Comment by esizer** [comment]
> To make a plan to go forward, what would you call the main blockers then for moving forward. As in, of the things you noted what should be spun off as issues marked as blocking this. If we can't progress on this issue itself we can at least plan the next steps

I think this is the main one since we need to eventually bump urql to v4.

 - https://github.com/GCTC-NTGC/gc-digital-talent/issues/7887

Theses ones are less important but may make working with the graphcache more straightforward (i think but not sure 😅 )

 - https://github.com/GCTC-NTGC/gc-digital-talent/issues/4662
 - https://github.com/GCTC-NTGC/gc-digital-talent/issues/6726
 - https://github.com/GCTC-NTGC/gc-digital-talent/issues/8401

**Comment by vd1992** [comment]
Marked the first issue as a blocker. We should consider tackling the second finally someday, been TODO for a while and is blocking multiple issues in the product backlog

**Comment by esizer** [comment]
Based on our discussion in stand-up about not wanting to deal with the graphcache right now, I will be putting this down again 😅 .

If anyone is interested, I did get reasonably far and have published by branch if your interested in the initial work and want to pick it up!

https://github.com/GCTC-NTGC/gc-digital-talent/tree/8309-urql-optimistic-updates

**Comment by tristan-orourke** [comment]
Most of the discussion on this ticket is about optimistic updates. I'll rewrite the issue to reflect that, and move the original, more general, issue into a new ticket. #9515

**Comment by esizer** [comment]
Okay, last time I tried this I ran into some issues but I think I know what they are and we have some options to try and fix them. 

## Issue

With graphcache, it only really "fetches" an entity once (if it does not exist in the cache). Once something is in the cache with a matching `id` and `__typename`, it no longer counts as "fetching" and just a cache update, even if it makes a trip to the server.

### Where it is going wrong

Right now, we have the `Pending` component that shows a spinner based on a boolean prop (`fetching`) then loads its children once that boolean is false. 

Our pattern right now is to check that data exists, if so we render the actual page component. Otherwise, we throw a 404 and the error page is rendered.

Since `fetching` is false during a partial update, we try to render the children too early (before data exists) then we either see an empty page or 404.

## Solution

I think all we need to do is check to see if data exists before anything, so just add it to the `fetching` check in our pending component. However, this leads to pop in since data cane be partial. So, there is also `stale` that appears when the cache is incomplete or stale and it is sending a request to grab the latest data.

### Example

```tsx
const Component = () {
  const [{ data, fetching, stale, error }] = useQuery({ ... });

  return (
    <Pending fetching={fetching || stale || !data} error={error}>
      {data?.entity ? <Page query={data.entity} /> : <ThrowNotFound />}
    </Pending>
  );
}
```

### Experiment

https://github.com/GCTC-NTGC/gc-digital-talent/tree/8309-urql-graphcache

I have created a branch that sets up grapahcache and fixed the issues on the `ProfileAndApplicationsPage` and `CareerTimelineAndRecruitmentPage`.

See what I mean by pulling the branch, building and following these steps with a logged in user.

1. Carefully observe the pop in on user pages
2. Add `stale` check and see the loading spinner stays slightly longer but no pop in anymore
3. Navigate to a skill page `/applicant/skills/{skillId}` (that does not check for data)
4. Observe 404
5. Add data and stale check
6. Confirm no more unexpected 404

**Comment by tristan-orourke** [comment]
Are there any mutations which modify more than the Type they return?

**Comment by esizer** [comment]
Closing this but keeping branch around so @tristan-orourke can take a look at it when he has time.
